content/automate/ManageIQ/Transformation/StateMachines/VMTransformation.class/__methods__/update_status.rb

```diff
@@ -0,0 +1,119 @@
+module Transformation
+  module StateMachines
+    module VMTransformation
+      class UpdateStatus
+        def initialize(handle = $evm)
+          @handle = handle
+        end
+
+        # This method gets the current state, even in the context of nested
+        # state machines. Where running a nested state machines, there can be
+        # more than one active states: one per level of nesting. The real
+        # active state is the one with the deepest ancestry.
+        def current_state(states)
+          active_states = states.select { |_, v| v['status'] == 'active' }.keys
+          states_depth = active_states.map { |state| state.split('/').length }
+          active_states[states_depth.index(states_depth.max)]
+        end
+
+        # This method recusrively collects the progress of children states
+        # and weight them, to finally weight the result with the current state
+        # weight.
+        def reconcile_children_percent(path, states)
+          children = states.reject { |k, _| k.gsub(/^#{path}\//, '').include?('/') }
+          if children.empty?
+            percent = states[path]['percent'].to_f * states[path]['weight'].to_f / 100.to_f
+          else
+            percent = 0
+            children.each_key { |child| percent += reconcile_children_percent(child, states).to_f * children[child]['weight'].to_f / 100.to_f }
+          end
+        end
+
+        def main
+          task = @handle.root['service_template_transformation_plan_task']
+          unless task.nil?
+            # Initiate the task progress hash if it doesn't exist yet
+            progress = task.get_option(:progress) || { 'current_state' => '', 'current_description' => '', 'percent' => 0, 'states' => {} }
+
+            # Collect the state details
+            state_name = @handle.root['ae_state']
+            state_ancestry = @handle.object['state_ancestry'].to_s
+            state_description = @handle.inputs['state_description'] == '_' ? state_name : @handle.inputs['state_description']
+            state_weight = @handle.inputs['state_weight']
+
+            # Get the current state hash if it already exists
+            state_hash = progress['states']["#{state_ancestry}/#{state_name}"]
+
+            # Get the state progress data provided by the state method
+            state_progress = @handle.get_state_var('ae_state_progress')
+            
+            # Aggregate the state progress in the state hash based on state
+            # phase (entry, exit, error). This updates the details with live
+            # data from the method, and also sets sane defaults.
+            case @handle.root['ae_status_state']
+            when 'on_entry'
```

Comment:
This `main` method should be broken down into smaller methods.  A good place to start is converting these `when` blocks for the case statement into separate methods.